### ISSUE TITLE
An improved asm command

### ISSUE THREAD

**Issue by kinnala** [kinnala/scikit-fem#15]
I was doing some benchmarks on assembling larger problems. I noticed that there is some unnecessary overhead in the approach used [here](https://github.com/kinnala/scikit-fem/blob/482ffe35e095ad8ef3c96fd695e9c9e51cbe5428/skfem/assembly.py#L439-L442).

The overhead is not visible in smaller problems but starts to be significant in 3D after 2M DOF's or so.

Some remarks:
- Assembling similar matrices in MATLAB using *sparse* is faster
- It seems that this was solved in SfePy by writing a [separate matrix assembler in Cython](https://github.com/sfepy/sfepy/blob/master/sfepy/discrete/common/extmods/assemble.pyx)

**Comment by gdmcbain** [comment]
Could you share the benchmark?  I've also been doing some profiling here and it might help if our measures are commensurate.

For mine so far (2D, 1.5M DOFs), cProfile revealed that almost half of the time was spent in `meshio.read`, so I was going to look into that before assembly or solution.

Before the profiling, I had thought that assembly might be the culprit and had been thinking about Cython or [Numba](http://numba.pydata.org).  After reading [`The case for Numba in community code`](https://www.anaconda.com/blog/developer-blog/the-case-for-numba-in-community-code) I thought that an advantage of the latter in the context of scikit-fem might be that in happy cases it only involves a wrapping a [`@jit`](http://numba.pydata.org/numba-doc/latest/user/jit.html) decorator around native Python which can be ignored if Numba isn't available.  As my immediate problem lay in I/O, I didn't measure this yet.

**Comment by kinnala** [comment]
I am not sure how to make a good benchmark as you would probably need to compare to MATLAB or SfePy.

You can see the effect in the following snippet. Save this to benchmark.py.

```python
from skfem import *                                                             
                                                                                
@bilinear_form                                                                  
def mass(u, du, v, dv, w):                                                      
    return u*v                                                                  
                                                                                
m = MeshTet()                                                                   
m.refine(6)                                                                     
ib = InteriorBasis(m, ElementTetP1(), MappingAffine(m), 2)                      
                                                                                
def run1():                                                                     
    return asm(mass, ib)                                                        
                                                                                
def run2():                                                                     
    return asm(mass, ib, assemble=False)   
```

Then in iPython do

```
%run benchmark.py
%time run1()
%time run2()
```

and compare the timings. I suppose the difference is larger than it should be.

I think one could do something similar to what is done in SfePy using Numba.

**Comment by kinnala** [comment]
I forgot to add my timings

```
In [1]: %run benchmark.py

In [2]: %time run1()
CPU times: user 28.2 s, sys: 1.56 s, total: 29.8 s
Wall time: 29.9 s
Out[2]: 
<230945x230945 sparse matrix of type '<class 'numpy.float64'>'
	with 3363425 stored elements in Compressed Sparse Row format>

In [3]: %time run2()
CPU times: user 3.61 s, sys: 868 ms, total: 4.48 s
Wall time: 4.49 s
```
I suggest a modification to [this part](https://github.com/kinnala/scikit-fem/blob/482ffe35e095ad8ef3c96fd695e9c9e51cbe5428/skfem/assembly.py#L439-L442) so that the user can provide a function which inputs the variables _data_ (flattened), _rows_ and _cols_, and outputs a sparse matrix in any suitable format. A default implementation could be the current one, and we could then implement a faster alternative using Numba.

I can foresee a third alternative implementation to be handy in, e.g., GPU computations where the output could be a sparse array allocated on the GPU.

**Comment by gdmcbain** [comment]
Thank you.

Yes, after reading ['The case for Numba in community code'](https://www.anaconda.com/blog/developer-blog/the-case-for-numba-in-community-code) I was hoping Numba might do; compared to Cython, it seemed closer to the spirit of scikit-fem being native Python.

I don't have MATLAB but do have GNU Octave and some in-house finite element code which uses [`sparse`](https://octave.org/doc/interpreter/Creating-Sparse-Matrices.html#XREFsparse) to assemble finite-element bilinear forms.  I don't think I've run the latter since GNU Octave acquired a [JIT compiler](https://octave.org/doc/interpreter/JIT-Compiler.html#JIT-Compiler).  I also haven't installed SfePy since I switched to Python 3 - it's not nearly as easy  `pip install scikit-fem` .

**Comment by gdmcbain** [comment]
Wow, I see now why I hadn't noticed this issue before.  I'd been working on planar problems and it really is a three-dimensional issue, as you say.  Much bigger two-dimensional operators are much quicker; e.g.

```python
m = MeshTri()                                                               
m.refine(10)
```

gives 

```
%time run1()
CPU times: user 3.98 s, sys: 1.26 s, total: 5.24 s
Wall time: 5.25 s
Out[3]: 
<1050625x1050625 sparse matrix of type '<class 'numpy.float64'>'
	with 7346177 stored elements in Compressed Sparse Row format>
```

(This is on a slightly slower machine; taking about 40 s for the three-dimensional run1 cf. the 28 quoted above.)

**Comment by gdmcbain** [comment]
Hmm, hang on.  I'm getting quite different behaviour in comparing `run1` and `run2`.  I'm not that familiar with ipython so am running with your `benchmark.py` as copied and pasted into a file, then two separate other scripts, `benchmark1.py`:

```shell
from benchmark import run1

run1()
```

and `benchmark2.py`:

```shell
from benchmark import run2

run2()
```

Then running these with [`cProfile`](https://docs.python.org/3/library/profile.html#module-cProfile); e.g.
```shell
python -m cProfile -s cumtime benchmark2.py
```

The results are almost the same: 46 and 44 s.  Two thirds of this in each case is 30 s in `refine`.  The two calls to asm are 7 s and 4 s; so `assemble=False` makes a little bit of difference but not much.

Is ipython doing something clever like lazy evaluation or memoization?

**Comment by gdmcbain** [comment]
Actually no, I'm getting the same 7 and 4 s results in ipython too…; viz.

```
In [1]: %run benchmark.py

In [2]: %time run1()
CPU times: user 5.54 s, sys: 1.52 s, total: 7.06 s
Wall time: 7.08 s
Out[2]: 
<230945x230945 sparse matrix of type '<class 'numpy.float64'>'
	with 3363425 stored elements in Compressed Sparse Row format>

In [3]: %time run2()
CPU times: user 3.09 s, sys: 942 ms, total: 4.04 s
Wall time: 4.05 s
```

So I can't reproduce your 30 second run1 v. 4 second run2.  I'm running ipython 6.4.0 and Python 3.6.6 under Debian GNU/Linux testing (buster).

**Comment by kinnala** [comment]
It seems that this was fixed [already in SciPy 1.0.0](https://docs.scipy.org/doc/scipy/reference/release.1.0.0.html#scipy-sparse-improvements). For some reason I had an ancient version on my personal laptop.

I'm leaving the issue open since I still wish to make the [modifications discussed earlier](https://github.com/kinnala/scikit-fem/issues/15#issuecomment-401906816).

Thank you @gdmcbain for your input.

**Comment by kinnala** [comment]
I'm postponing this idea of modifying asm so that a practical use case appears.
